### PR TITLE
add BorderArray

### PR DESCRIPTION
**Pull request by jw3126**
#98

### CODE REVIEW

**Comment by jw3126** [comment]
@timholy what do you think about the general design of `BorderArray`?
* I did not implement `setindex!` since it would be mutating border + inner points simultaneously
* What should happen with `padarray`? Should it stay as is? Should it return a `BorderArray`? Should it be removed? Personally I think it could be return a `BorderArray` and the current `padarray` should be spelled something like `full(padarray(...))` instead.

**Comment by timholy** [comment]
This looks superficially awesome, but I haven't had bandwidth recently. Maybe I can review this tomorrow, sorry for the delay.

File: src/borderarray.jl [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        if !compatible_dimensions(arr, border)
+            throw(ArgumentError("$border lacks the proper padding sizes for an array with $(ndims(arr)) dimensions"))
+        end
+        A = typeof(arr)
+        B = typeof(border)
+        new{T,N,A,B}(arr, border)
+    end
+end
+
+# these are adaped from OffsetArrays
```
Comment (timholy) [inline src/borderarray.jl]:
```suggestion
# these are adapted from OffsetArrays
```

File: src/borderarray.jl [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    ci = CartesianIndex(inds)
+    o[ci]
+end
+
+@inline function Base.getindex(o::BorderArray, ci::CartesianIndex)
+    if checkbounds(Bool, o.inner, ci)
+        @inbounds o.inner[ci]
+    else
+        @boundscheck checkbounds(o, ci)
+        getindex_outer_inbounds(o.inner, o.border, ci)
```
Comment (timholy) [inline src/borderarray.jl]: Do we need to make sure that both paths return values of type `T`? AFAICT it's possible to do something like
```julia
v = BorderArray([1, 2], Fill(NaN, (1,)))
```
and then indexing `v` is non-inferrable.

File: src/border.jl [hunk 3]
```diff
@@ -652,13 +652,11 @@ The command `padarray(A,Fill(0,(1,1,1)))` yields
...
-    inds = padindices(img, border)
-    # like img[inds...] except that we can control the element type
-    newinds = map(Base.axes1, inds)
-    dest = similar(img, T, newinds)
-    copydata!(dest, img, inds)
+padarray(img::AbstractArray, border::AbstractBorder)  = padarray(eltype(img), img, border)
+function padarray(::Type{T}, img::AbstractArray, border) where {T}
+    ba = BorderArray(img, border)
+    out = similar(ba, T)
+    copy!(out, ba)
```
Comment (timholy) [inline src/border.jl]: Can you do some performance comparisons between the existing implementation and this one? Your performance optimizations for `_copy!` are encouraging, of course, but just want to check.

Comment (jw3126) [reply src/border.jl]: Yes I did. The current implementation is basically stolen from `padarray` anyway. I tinkered a bit with making it simpler or using the generic `copy`, but sadly performance was not as good.

File: src/borderarray.jl (hunk 2 above)
Comment (jw3126) [reply src/borderarray.jl]: Good point! Do you prefer to throw an error in the constructor or a `convert` in `getindex` if that happens?

Comment (timholy) [reply src/borderarray.jl]: If it's straightforward to do it in the constructor, that would be best.

**Comment by timholy** [comment]
Once this passes tests, I approve merging. Thanks for this, fantastic work!

**Comment by jw3126** [comment]
@timholy can we tag a new release?

**Comment by timholy** [comment]
Sorry I didn't notice the question about tagging. Absolutely! There's nothing breaking here, right? So this could be 0.6.1?

I think you should be able tag yourself, if you like. Perhaps give it a try, just so we know whether you are as empowered as you deserve to be.

**Comment by jw3126** [comment]
Yeah, there is no breaking change, at least no intended one :smile:. I will see if I can create a release.

**Pull request merged**